src/lem/mod.rs

```diff
@@ -422,9 +598,327 @@ mod tests {
+                            let expr_comm3: Cons = hash2(expr_comm, expr_comm2);
+                            let expr_comm4: Cons = hash2(expr_comm, expr_comm3);
+                            let expr_comm5: Cons = hash2(expr_comm, expr_comm4);
+                        },
+                        Outermost => {
+                            let expr_outer: Cons = hash2(expr_in, expr_in);
+                            let expr_outer2: Cons = hash2(expr_outer, expr_outer);
+                        }
+                    };
+                    let expr_aux: Cons = hash2(expr_in, expr_in);
+                    let expr_out: Cons = hash2(expr_aux, expr_aux);
+                    let cont_out_terminal: Terminal;
+                    return (expr_out, env_in, cont_out_terminal);
+                },
+                Char => {
+                    let expr_aux: Cons = hash2(expr_in, expr_in);
+                    let expr_aux2: Cons = hash2(expr_aux, expr_aux);
+                    let expr_out: Cons = hash2(expr_aux2, expr_aux2);
+                    let cont_out_terminal: Terminal;
+                    return (expr_out, env_in, cont_out_terminal);
+                }
+            };
+        })
+        .unwrap();
+
+        constrain_test_helper(&lem, &Ptr::num(Fr::from_u64(42)), NumSlots::new((7, 0, 0)));
+    }
+
+    #[test]
+    fn test_hash2_slots_seq() {
+        let lem = lem!(expr_in env_in cont_in {
+            let expr_aux: Cons = hash2(expr_in, expr_in);
+            let expr_out: Cons = hash2(expr_aux, expr_aux);
+            let cont_out_terminal: Terminal;
+            return (expr_out, env_in, cont_out_terminal);
+        })
+        .unwrap();
+
+        constrain_test_helper(&lem, &Ptr::num(Fr::from_u64(42)), NumSlots::new((2, 0, 0)));
+    }
+
+    #[test]
+    fn test_hash3_slots_simple() {
+        let lem = lem!(expr_in env_in cont_in {
+            match_tag expr_in {
+                Num => {
+                    let expr_out: Cons = hash3(expr_in, expr_in, expr_in);
+                    let cont_out_terminal: Terminal;
+                    return (expr_out, env_in, cont_out_terminal);
+                }
+            };
+        })
+        .unwrap();
+
+        constrain_test_helper(&lem, &Ptr::num(Fr::from_u64(42)), NumSlots::new((0, 1, 0)));
+    }
+
+    #[test]
+    fn test_hash3_slots_unhash_simple() {
+        let lem = lem!(expr_in env_in cont_in {
+            match_tag expr_in {
+                Num => {
+                    let expr_out: Cons = hash3(expr_in, expr_in, expr_in);
+                    let (expr_input1, expr_input2, expr_input3) = unhash3(expr_out);
+                    let cont_out_terminal: Terminal;
+                    return (expr_input1, env_in, cont_out_terminal);
+                }
+            };
+        })
+        .unwrap();
+
+        constrain_test_helper(&lem, &Ptr::num(Fr::from_u64(42)), NumSlots::new((0, 2, 0)));
```

Comment:
When hash deduplication is implemented, this should become `(0, 1, 0)` — if I'm following — since the above will effectively prove the same hash twice.